components/builder-jobsrv/src/server/log_ingester.rs

```diff
@@ -0,0 +1,142 @@
+            //    L = a line of log output
+            //    C = the log is complete
+            // 3: a protobuf message
+
+            self.intake_sock.recv(&mut self.msg, 0)?; // identity frame
+
+            match str::from_utf8(self.intake_sock.recv_bytes(0).unwrap().as_slice()).unwrap() {
+                "L" => {
+                    self.intake_sock.recv(&mut self.msg, 0)?; // protobuf message frame
+                    match parse_from_bytes::<JobLogChunk>(&self.msg) {
+                        Ok(chunk) => {
+                            let log_file = self.log_dir.log_file_path(chunk.get_job_id());
+
+                            // TODO: Consider caching file handles for
+                            // currently-processing logs.
+                            let open = OpenOptions::new()
+                                .create(true)
+                                .append(true)
+                                .open(log_file.as_path());
+
+                            match open {
+                                Ok(mut file) => {
+                                    // TODO: Just write the chunk body; line number is
+                                    // for debugging only
+                                    file.write(format!("{}: {}",
+                                                       chunk.get_seq(),
+                                                       chunk.get_body()).as_bytes())?;
+                                    file.flush()?;
+                                    println!("LOG: {:?}: {:?}",
+                                             chunk.get_seq(),
+                                             chunk.get_body());
+                                },
+                                Err(e) => {
+                                    warn!("Could not open {:?} for appending! {:?}", log_file, e);
+                                }
+                            }
+                        },
+                        Err(e) => {
+                            warn!("ERROR parsing JobLogChunk: {:?}", e);
+                        }
+                    }
+                },
+                "C" => {
+                    self.intake_sock.recv(&mut self.msg, 0)?; // protobuf message frame
+                    match parse_from_bytes::<JobLogComplete>(&self.msg) {
+                        Ok(complete) => {
+                            let id = complete.get_job_id();
+                            debug!("Log complete for job {:?}", id);
+
+                            // TODO: Save logs in long-term storage,
+                            // like S3
+                            let log_file = self.log_dir.log_file_path(id);
+
+                            // Until we can ship things off to an
+                            // S3-alike, we'll just use a file URL
+                            let url = format!("file://{}", log_file.to_str().unwrap());
+                            self.data_store.set_log_url(id, &url)?;
```

Comment:
Once we have off-disk storage, we'll look to see if there's a log URL whenever someone asks for the logs; if it's there, we'll just retrieve it and return it to the user as a complete log. On the other hand, if the log data is still coming in, we don't want to be constantly updating and retrieving from an S3 bucket, so we'll just keep using the local file on disk until it's complete.

In this scenario, there's never going to be a time where we have a log stored off-disk (and thus need a URL) but are also still receiving additional lines to append to it, so I think having an additional flag wouldn't add anything.